File: frontend/src/types/survey.ts
export type QuestionType = 'multiple_choice' | 'single_choice' | 'text' | 'textarea' | 'rating_5' | 'rating_10' | 'yes_no';
export type SurveyStatus = 'draft' | 'active' | 'paused' | 'completed' | 'archived';
export type SurveyAccessType = 'invite_only' | 'public';
export type InvitationStatus = 'pending' | 'sent' | 'viewed' | 'started' | 'completed' | 'expired';

export interface QuestionOption {
  id: string;
  text: string;
  value: string | number;
}

export interface SurveyQuestion {
  id: string;
  type: QuestionType;
  text: string;
  description?: string;
  required: boolean;
  options?: QuestionOption[];
  min_rating?: number;
  max_rating?: number;
  order: number;
}

export interface TargetSegment {
  tier_restrictions?: string[];
  min_points?: number;
  max_points?: number;
  registration_after?: string;
  registration_before?: string;
  oauth_providers?: string[];
  exclude_users?: string[];
}

export interface Survey {
  id: string;
  title: string;
  description?: string;
  questions: SurveyQuestion[];
  target_segment: TargetSegment;
  status: SurveyStatus;
  access_type: SurveyAccessType;
  scheduled_start?: string;
  scheduled_end?: string;
  created_by?: string;
  created_at: string;
  updated_at: string;
}

export interface SurveyResponse {
  id: string;
  survey_id: string;
  user_id: string;
  answers: Record<string, any>;
  is_completed: boolean;
  progress: number;
  started_at: string;
  completed_at?: string;
  created_at: string;
  updated_at: string;
  user_email?: string;
  user_first_name?: string;
  user_last_name?: string;
}

export interface SurveyInvitation {
  id: string;
  survey_id: string;
  user_id: string;
  status: InvitationStatus;
  sent_at?: string;
  viewed_at?: string;
  expires_at?: string;
  created_at: string;
  updated_at: string;
}

// API Request/Response types
export interface CreateSurveyRequest {
  title: string;
  description?: string;
  questions: SurveyQuestion[];
  target_segment?: TargetSegment;
  access_type: SurveyAccessType;
  status?: SurveyStatus;
  scheduled_start?: string;
  scheduled_end?: string;
}

export interface UpdateSurveyRequest extends Partial<CreateSurveyRequest> {
  status?: SurveyStatus;
}

export interface SubmitResponseRequest {
  survey_id: string;
  answers: Record<string, any>;
  is_completed?: boolean;
}

export interface SurveyListResponse {
  surveys: Survey[];
  pagination: {
    total: number;
    page: number;
    limit: number;
    totalPages: number;
  };
}

export interface SurveyResponseListResponse {
  responses: SurveyResponse[];
  pagination: {
    total: number;
    page: number;
    limit: number;
    totalPages: number;
  };
}

export interface SurveyAnalytics {
  survey: Survey;
  responses: SurveyResponse[];
  totalResponses: number;
  completionRate: number;
  averageCompletionTime: number;
  responsesByDate: { date: string; count: number }[];
  questionAnalytics: {
    questionId: string;
    question: string;
    type: string;
    responses: Record<string, number>;
    averageRating?: number;
  }[];
}

// Survey Coupon Assignment types
// Note: Coupons are always awarded on survey completion

export interface SurveyCouponAssignment {
  id: string;
  survey_id: string;
  coupon_id: string;
  is_active: boolean;
  max_awards?: number;
  awarded_count: number;
  assigned_by?: string;
  assigned_reason: string;
  custom_expiry_days?: number;
  created_at: string;
  updated_at: string;
}

export interface SurveyCouponDetails {
  assignment_id: string;
  survey_id: string;
  survey_title: string;
  survey_status: string;
  coupon_id: string;
  coupon_code: string;
  coupon_name: string;
  coupon_type: string;
  coupon_value?: number;
  coupon_currency: string;
  coupon_status: string;
  is_active: boolean;
  max_awards?: number;
  awarded_count: number;
  custom_expiry_days?: number;
  assigned_reason: string;
  assigned_by?: string;
  assigned_by_email?: string;
  assigned_at: string;
  updated_at: string;
}

export interface SurveyRewardHistory {
  id: string;
  survey_coupon_assignment_id: string;
  survey_response_id: string;
  user_coupon_id: string;
  user_id: string;
  awarded_at: string;
  award_condition_met: string;
  metadata: Record<string, any>;
  created_at: string;
  // Additional display fields
  user_email?: string;
  user_name?: string;
  coupon_code?: string;
  coupon_name?: string;
}

// API Request types for survey-coupon assignments
export interface AssignCouponToSurveyRequest {
  survey_id: string;
  coupon_id: string;
  max_awards?: number;
  custom_expiry_days?: number;
  assigned_reason?: string;
}

export interface UpdateSurveyCouponAssignmentRequest {
  max_awards?: number;
  custom_expiry_days?: number;
  assigned_reason?: string;
  is_active?: boolean;
}

export interface SurveyCouponAssignmentListResponse {
  assignments: SurveyCouponDetails[];
  total: number;
  page: number;
  limit: number;
  totalPages: number;
}

export interface SurveyRewardHistoryResponse {
  rewards: SurveyRewardHistory[];
  total: number;
  totalPages: number;
}

File: backend-rust/src/models/survey.rs
//! Survey models
//!
//! Contains structs for surveys, survey questions, and survey responses.

use chrono::{DateTime, NaiveDateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Survey database entity
#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct Survey {
    pub id: Uuid,
    pub title: String,
    pub description: Option<String>,
    pub questions: serde_json::Value,
    pub target_segment: Option<serde_json::Value>,
    pub status: Option<String>,
    pub scheduled_start: Option<NaiveDateTime>,
    pub scheduled_end: Option<NaiveDateTime>,
    pub created_by: Option<Uuid>,
    pub created_at: Option<NaiveDateTime>,
    pub updated_at: Option<NaiveDateTime>,
    pub access_type: String,
    pub original_language: Option<String>,
    pub available_languages: Option<serde_json::Value>,
    pub last_translated: Option<DateTime<Utc>>,
    pub translation_status: Option<String>,
}

/// Survey question structure (stored as JSON in database)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SurveyQuestion {
    pub id: String,
    pub question_type: SurveyQuestionType,
    pub text: String,
    pub description: Option<String>,
    pub required: bool,
    pub options: Option<Vec<SurveyQuestionOption>>,
    pub validation: Option<SurveyQuestionValidation>,
    pub order: i32,
}

/// Survey question type
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SurveyQuestionType {
    SingleChoice,
    MultipleChoice,
    Text,
    TextArea,
    Rating,
    Scale,
    Date,
    Number,
    Email,
    Phone,
}

/// Survey question option (for choice-based questions)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SurveyQuestionOption {
    pub id: String,
    pub text: String,
    pub value: Option<String>,
    pub order: Option<i32>,
}

/// Survey question validation rules
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SurveyQuestionValidation {
    pub min_length: Option<i32>,
    pub max_length: Option<i32>,
    pub min_value: Option<f64>,
    pub max_value: Option<f64>,
    pub pattern: Option<String>,
    pub custom_error: Option<String>,
}

/// Survey response database entity
#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct SurveyResponse {
    pub id: Uuid,
    pub survey_id: Option<Uuid>,
    pub user_id: Option<Uuid>,
    pub answers: serde_json::Value,
    pub is_completed: Option<bool>,
    pub progress: Option<i32>,
    pub started_at: Option<NaiveDateTime>,
    pub completed_at: Option<NaiveDateTime>,
    pub created_at: Option<NaiveDateTime>,
    pub updated_at: Option<NaiveDateTime>,
}

/// Survey invitation database entity
#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct SurveyInvitation {
    pub id: Uuid,
    pub survey_id: Option<Uuid>,
    pub user_id: Option<Uuid>,
    pub status: Option<String>,
    pub sent_at: Option<NaiveDateTime>,
    pub viewed_at: Option<NaiveDateTime>,
    pub expires_at: Option<NaiveDateTime>,
    pub created_at: Option<NaiveDateTime>,
    pub updated_at: Option<NaiveDateTime>,
}

/// Create survey request DTO
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateSurveyRequest {
    pub title: String,
    pub description: Option<String>,
    pub questions: Vec<SurveyQuestion>,
    pub target_segment: Option<serde_json::Value>,
    pub status: Option<String>,
    pub scheduled_start: Option<NaiveDateTime>,
    pub scheduled_end: Option<NaiveDateTime>,
    pub access_type: Option<String>,
}

/// Update survey request DTO
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateSurveyRequest {
    pub title: Option<String>,
    pub description: Option<String>,
    pub questions: Option<Vec<SurveyQuestion>>,
    pub target_segment: Option<serde_json::Value>,
    pub status: Option<String>,
    pub scheduled_start: Option<NaiveDateTime>,
    pub scheduled_end: Option<NaiveDateTime>,
    pub access_type: Option<String>,
}

/// Survey response DTO
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SurveyResponseDto {
    pub id: Uuid,
    pub title: String,
    pub description: Option<String>,
    pub questions: Vec<SurveyQuestion>,
    pub status: Option<String>,
    pub scheduled_start: Option<NaiveDateTime>,
    pub scheduled_end: Option<NaiveDateTime>,
    pub access_type: String,
    pub created_at: Option<NaiveDateTime>,
}

/// Submit survey response request DTO
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SubmitSurveyResponseRequest {
    pub survey_id: Uuid,
    pub answers: serde_json::Value,
    pub is_completed: bool,
}

/// Survey response answer DTO
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SurveyAnswerDto {
    pub id: Uuid,
    pub survey_id: Option<Uuid>,
    pub user_id: Option<Uuid>,
    pub answers: serde_json::Value,
    pub is_completed: Option<bool>,
    pub progress: Option<i32>,
    pub started_at: Option<NaiveDateTime>,
    pub completed_at: Option<NaiveDateTime>,
}

impl From<Survey> for SurveyResponseDto {
    fn from(survey: Survey) -> Self {
        let questions: Vec<SurveyQuestion> =
            serde_json::from_value(survey.questions.clone()).unwrap_or_default();

        Self {
            id: survey.id,
            title: survey.title,
            description: survey.description,
            questions,
            status: survey.status,
            scheduled_start: survey.scheduled_start,
            scheduled_end: survey.scheduled_end,
            access_type: survey.access_type,
            created_at: survey.created_at,
        }
    }
}

impl From<SurveyResponse> for SurveyAnswerDto {
    fn from(response: SurveyResponse) -> Self {
        Self {
            id: response.id,
            survey_id: response.survey_id,
            user_id: response.user_id,
            answers: response.answers,
            is_completed: response.is_completed,
            progress: response.progress,
            started_at: response.started_at,
            completed_at: response.completed_at,
        }
    }
}
